File: src/backend/GoCalGo.Contracts/Events/PokemonRole.cs
using System.Text.Json.Serialization;

namespace GoCalGo.Contracts.Events
{
    /// <summary>
    /// Describes a Pokemon's role within an event context.
    /// Serialised as lowercase kebab-case strings.
    /// </summary>
    [JsonConverter(typeof(JsonStringEnumConverter<PokemonRole>))]
    public enum PokemonRole
    {
        [JsonStringEnumMemberName("spawn")]
        Spawn,

        [JsonStringEnumMemberName("shiny")]
        Shiny,

        [JsonStringEnumMemberName("spotlight")]
        Spotlight,

        [JsonStringEnumMemberName("raid-boss")]
        RaidBoss,

        [JsonStringEnumMemberName("research-reward")]
        ResearchReward,

        [JsonStringEnumMemberName("research-breakthrough")]
        ResearchBreakthrough,
    }
}

File: src/app/lib/models/pokemon_role.dart
/// Describes a Pokemon's role within an event context.
enum PokemonRole {
  spawn('spawn'),
  shiny('shiny'),
  spotlight('spotlight'),
  raidBoss('raid-boss'),
  researchReward('research-reward'),
  researchBreakthrough('research-breakthrough');

  const PokemonRole(this.value);
  final String value;

  static PokemonRole fromJson(String json) =>
      PokemonRole.values.firstWhere(
        (e) => e.value == json,
        orElse: () => PokemonRole.spawn,
      );

  String toJson() => value;
}
